woocommerce/woocommerce-ios
# Update `ListSelectorViewController` for usage in products tab filtering

Prerequisites for #2037 

## Background

In #2037, I'm planning to implement the UI by splitting the UI into 2 container views in a vertical stack view:
- the top view contains a `UINavigationController` with a `ListSelectorViewController` with filter rows and their values
  - Tapping on a filter row navigates to another `ListSelectorViewController` with the options
- the bottom view contains a CTA to filter the products with the given filters

<img width="443" alt="Screen Shot 2020-04-27 at 4 32 44 PM" src="https://user-images.githubusercontent.com/1945542/80351763-71e47380-88a5-11ea-8ee6-275b71a92569.png">

In order for me to implement the filter Products UI/UX using `ListSelectorViewController` for all the list selector screens (including the first screen), there are a few changes to be made as listed in the next section along with some renaming.


## Changes

- Renamed `ListSelectorDataSource` to `ListSelectorCommand` to be more consistent with naming based on the discussion in the bottom sheet list selector PR https://github.com/woocommerce/woocommerce-ios/pull/2155#discussion_r411438712
- In `ListSelectorViewController`'s `viewWillDisappear`, added `isMovingFromParent` check before calling `onDismiss(command.selected)` thanks to @shiki's suggestion in https://github.com/woocommerce/woocommerce-ios/pull/2189#discussion_r414853424
- Moved `navigationBarTitle` from `ListSelectorViewProperties` to `ListSelectorCommand`, and removed `ListSelectorViewProperties`
  - In #2037, the navigation bar title is no longer constant and I thought it could also simplify the command by not having a separate struct for UI properties. If we feel like separating the UI properties, we could create another protocol and have the command extend the UI properties protocol
- In `ListSelectorDataSource`, passed the list selector view controller in `handleSelectedChange`
  - In #2037, we are navigating to another filter list selector when the user taps on a filter row and I felt like this is the most appropriate place to let the command implementation to know about the view controller
- In `ListSelectorViewController`, added a function to reload data
  - In #2037, we have to refresh the UI when any filter value changes other than when the selected model changes. The UI changes include the navigation bar title and table view
- Renamed `*ListSelectorDataSource` to `*ListSelectorCommand` and updated unit tests:
  - `ProductBackordersSettingListSelectorDataSource` → `ProductBackordersSettingListSelectorCommand`
  - `ProductStockStatusListSelectorDataSource` → `ProductStockStatusListSelectorCommand`
  - `ProductTaxStatusListSelectorDataSource` → `ProductTaxStatusListSelectorCommand`
  - `ProductStatusSettingListSelectorDataSource` → `ProductStatusSettingListSelectorCommand`

## Testing

No behavior changes are expected from this PR, and you can quickly go over the affected screens with the following steps:

- Go to the Products tab
- Tap on a simple Product
- Tap on "⋯" in the navigation bar
- Tap on "Product settings"
- Tap on "Status" row --> the product status selector UI/UX should be the same as before
- Go back to the Edit Product main screen
- Tap on the Price row
- Tap on the "Tax status" row --> the product tax status selector UI/UX should be the same as before
- Go back to the Edit Product main screen
- Tap on the Inventory row
- Enable the "Manage stock" setting
- Tap on the "Backorders" row --> the product backorders setting selector UI/UX should be the same as before
- Navigate back to the inventory settings
- Disable the "Manage stock" setting
- Tap on the "Stock status" row --> the product stock status selector UI/UX should be the same as before

## Example screenshots

Product stock status (`ProductStockStatusListSelectorCommand`) | Backorders (`ProductBackordersSettingListSelectorCommand`) | Product tax status (`ProductTaxStatusListSelectorDataSource`) | Product status (`ProductStatusSettingListSelectorDataSource`)
-- | -- | -- | --
![Simulator Screen Shot - iPhone 11 - 2020-04-27 at 15 50 39](https://user-images.githubusercontent.com/1945542/80352222-22eb0e00-88a6-11ea-9296-2f02e844fb22.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-27 at 15 50 44](https://user-images.githubusercontent.com/1945542/80352229-25e5fe80-88a6-11ea-8b2d-cab1eb143530.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-27 at 15 50 53](https://user-images.githubusercontent.com/1945542/80352232-267e9500-88a6-11ea-8b62-e8bdcf6e3004.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-27 at 15 51 25](https://user-images.githubusercontent.com/1945542/80352235-27172b80-88a6-11ea-9ee2-af3903d666d9.png)


Update release notes:
